docs/polymorphic-table-types.md

```diff
@@ -0,0 +1,198 @@
+	return 'Foo'
+end)
+
+-- <'Bar'>(() -> 'Bar') -> Factory<'Bar'> ==> Factory<'Bar'>
+local B = world.factory(function()
+	return 'Bar'
+end)
+
+-- number
+local e = world.entity()
+
+-- 'Foo'
+A.add(e)
+
+-- 'Bar'
+B.add(e)
+
+-- { [Factory<any>]: any? } 	(no possible way to extract the generic)
+local components = world.get(e)
+
+-- { [Factory<any>]: any? }[Factory<'Foo'>] ==> any?
+local a = components[A]
+
+-- { [Factory<any>]: any? }[Factory<'Bar'>] ==> any?
+local b = components[B]
+```
+
+We've lost our type information of `a` and `b` in the last table step!
+
+-   "Why not use a getter?"
+    -   Mentioned before, API bloat or performance concerns, getting components from this table usually happens during iterations and thousands of times a frame.
+-   "Why not implement better metamethod inference?"
+    -   The `components` table guarantees users may define their own metatables. Metatables are not an elegant, reliable, or wholly-acceptable solution to this lack of types. However better inference is always a good thing! This is a great idea for a _different_ RFC.
+
+## Design
+
+We propose naturally extending this type-mapping capability to tables in the backwards-compatible form `<T>{ [F<T>]: G<T> }`. We recycle our familiar syntax of polymorphic functions and use them in a very similar fashion. If one understands polymorphic functions, they will understand this as well. It is backwards compatible as it is currently invalid syntax.
+
+### General Indexer Solution
+
+To rewrite our previous code examples using the new type:
+
+```lua
+local foo = {} :: <T>{ [F<T>]: G<T> }
+```
+
+This is not only simpler code, but shorter without all the duplication or extra api that comes with such a structure. Looking at the syntax, it reads very similarly to if it were a function:
+
+```lua
+type A = <T>{ [F<T>]: G<T> }
+type B = <T>(F<T>) -> G<T>
+```
+
+### Specific Indexer Solution
+
+Back to the ECS example, we change the type of `components` to be what it should have been all along:
+
+```lua
+-- <C>{ [Factory<C>]: C? }		(look we extracted the generic!)
+local components = world.get(e)
+
+-- <'Foo'>{ [Factory<'Foo'>]: 'Foo'? }[Factory<'Foo'>] ==> 'Foo'?
+local a = components[A]
+
+-- <'Bar'>{ [Factory<'Bar'>]: 'Bar'? }[Factory<'Bar'>] ==> 'Bar'?
+local b = components[B]
+```
+
+And perfect intellisense is now achieved without a single typecast or getter (per component). It was simple too, encouraging the simplest solution in the process.
+
+We believe that this implementation is a natural step in the direction towards a fuller Luau type-system, as it already has a functional cousin and practical use-cases. Concerns with api coherency, stylistic coherency, learning curve, acceptance, and more already answered with polymorphic functions. If you want to understand the behavior of a polymorphic table, simply look to the behavior of polymorphic functions as they will have the exact same properties.
+
+`function` $\cong$ `table`
```

Comment:
This is a dangerous relationship to draw, because mutability changes the rules _dramatically._ Functions, being immutable, have broader subtyping rules - contravariant for arguments and covariant for returns - than mutable tables do. In order to remain type-safe in the face of mutability, table fields must be immutable - they currently aren't always treated like this, but we're working towards that.